File: 06/main.cpp
#include <algorithm>
#include <fstream>
#include <unordered_set>

using namespace std;

template <size_t N>
void populateArray(char (&memory)[N])
{
    for (size_t i = 0; i < N - 1; i++)
    {
        memory[i] = memory[N - 1];
    }
}

template <size_t N>
bool checkUnique(const char (&chars)[N])
{
    unordered_set<char> set(begin(chars), end(chars));
    return set.size() == N;
}

int main()
{
    ifstream inputStream;
    inputStream.open("input.txt");

    char inputChar;
    char memoryOne[4];
    char memoryTwo[14];

    bool partOneDone = false;
    bool partTwoDone = false;

    size_t index = 0;
    while (inputStream.get(inputChar) && !(partOneDone & partTwoDone))
    {
        index++;
        memoryOne[3] = inputChar;
        memoryTwo[13] = inputChar;
        if (index == 1)
        {
            populateArray(memoryOne);
            populateArray(memoryTwo);
        }

        if (checkUnique(memoryOne) & !partOneDone)
        {
            printf("Part one: %d\n", (int)index);
            partOneDone = true;
        }

        if (checkUnique(memoryTwo) & !partTwoDone)
        {
            printf("Part two: %d\n", (int)index);
            partTwoDone = true;
        }

        memoryOne[index % 3] = memoryOne[3];
        memoryTwo[index % 13] = memoryTwo[13];
    }

    return 0;
}

File: 06/main.py
import pandas as pd;


def read_file(file_name):
    try:
        with open(file_name, "r") as file:
            content = file.read()
    except Exception as e:
        print("Something went wrong: ", e)

    return content


def process_data(input, window):
    rows = [list(input[i:i+window]) for i in range(len(input)-window)]

    return pd.DataFrame(rows)


def main():
    input = read_file('input.txt')
    
    size_1 = 4
    size_2 = 14

    data_1 = process_data(input, size_1)
    data_2 = process_data(input, size_2)

    result_1 = size_1 + data_1[data_1.nunique(axis=1) == size_1].head(1).index.values[0]
    result_2 = size_2 + data_2[data_2.nunique(axis=1) == size_2].head(1).index.values[0]

    print("Part one: {}\nPart two: {}".format(result_1, result_2))


if __name__ == "__main__":
    main()